### PR TITLE
Fix memory leak due to handle

### LINKED ISSUE #2184: Realm Memory Leak When Get Data.
## Goals
We receiver feed back from some user that host server is slow and disconnect many time. 
We see Ram Overload and we check detail, it maybe memory leak. 
We stop all function, test with just very simple get data from Ream for check bigger issue. 

## Expected Results

Memory is ok.

## Actual Results

Memory none stop Increasing. 

## Steps to Reproduce

We check memory leak, and we thing it is leaking. 
Test function by function to finding where it leak => We found it in very simple function.
Test many way but can't solve. 
Finding google and issue in Realm but can't find anything like that. 

## Code Sample

Calling Function: 
 
        [HttpPost ("{id}/Measures", Name = "PostSchedules")]
        public string PostMeasure(string id, int child,  [FromBody] object value, string token)
        {
            Database.GWTokenDB.GetTokenGW(id); 
            return "{\"Result\":\"Successful\"}"; ;
        }

Action function: ( In test case it will return null - So Clone() function don't run in this test)

        public static GatewayTokenModel GetTokenGW(string gatewayId)
        {
            using (var realm = Realm.GetInstance(Configuration))
            {
                try
                {
                    var data = realm.Find<GatewayTokenModel>(gatewayId);
                    if (data != null) return data.Clone();
                    else return null;
                }
                catch (Exception ex)
                {

                }
            }
            return null;
        }

## Version of Realm and Tooling

- Realm Object Server Version: ? 5.1.2
- Flavor:
  - [ ] Realm Cloud
  - [x] Self-Hosted
- Server OS & Version: ? CentOS 6. 
- Client SDK Version: http request.
- Client OS & Version: Window 10, request http to server.

Adding Image Test Memory Leak: 
image

Capture Information at end of test => Stop send data more and press GC Collect many times until memory don't decreasing more.
image

### PR DESCRIPTION
**Pull request by papafe**
Added freeing of the handle in `State` when disposing of realm.

Fixes #2184

### CODE REVIEW

File: Realm/Realm/Handles/SharedRealmHandle.cs [hunk 1]
```diff
@@ -46,6 +46,9 @@ private static class NativeMethods
             [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
             public unsafe delegate void OpenRealmCallback(IntPtr task_completion_source, IntPtr shared_realm, int error_code, byte* message_buf, IntPtr message_len);
 
+            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+            public delegate void FreeGCHandleCallback(IntPtr handle);
```
Comment (nirinchev) [inline Realm/Realm/Handles/SharedRealmHandle.cs:50]: I know this was the name used previously, but let's rename to something that leaks less implementation details. I'm thinking either `DisposeStateCallback` or `OnBindingContextDestructedCallback`.

File: Realm/Realm/Handles/SharedRealmHandle.cs [hunk 2]
```diff
@@ -420,6 +425,15 @@ private static unsafe void HandleOpenRealmCallback(IntPtr taskCompletionSource,
             }
         }
 
+        [MonoPInvokeCallbackAttribute(typeof(NativeMethods.FreeGCHandleCallback))]
+        public static void FreeGCHandle(IntPtr handle)
+        {
+            if (handle != IntPtr.Zero)
+            {
+                GCHandle.FromIntPtr(handle).Free();
```
Comment (nirinchev) [inline Realm/Realm/Handles/SharedRealmHandle.cs:433]: Tied to my renaming suggestion - I'm not a big fan of this approach because it leaks the internal design of the managed state. Instead, what we could do is have `State` implement `IDisposable` and in the callback, we can cast the target to `State` and call the `Dispose` method.

**Review by nirinchev — COMMENTED** [review]
Is there a unit test we can write that would exhibit the memory leak? Maybe open a Realm and take a weak reference to the Realm and to its state, then run GC.Collect until the Realm gets collected and verify that the state is also collected?

**Comment by papafe** [comment]
> Is there a unit test we can write that would exhibit the memory leak? Maybe open a Realm and take a weak reference to the Realm and to its state, then run GC.Collect until the Realm gets collected and verify that the state is also collected?

At the moment the state is private, so to do that test I will need to change its accessibility at least to internal, and I don't feel it's worth it. I'll think about if we can do in a different way

**Comment by nirinchev** [comment]
You can cast to `IDisposable` and call the `Dispose` method.

**Review by LaPeste — APPROVED** [review]
An educational PR for me, since I haven't seen before the way C++ could be passed a callback from C#.
The unit test seems more than reasonable.
In summary I have no comments.

File: Realm/Realm/Realm.cs [hunk 3]
```diff
@@ -198,6 +198,7 @@ private static bool IsRealmOpen(string path)
 
         #endregion static
 
+        [SuppressMessage("Usage", "CA2213:Disposable fields should be disposed", Justification = "A realm is responsible for disposing of state as it could be shared")]
```
Comment (nirinchev) [inline Realm/Realm/Realm.cs:201]: This comment is a bit unclear - it says that "a realm is responsible for disposing", but it attempts to explain why the Realm will not dispose the state. I would suggest rewording it to something like "A State can be shared between multiple Realm instances. It is disposed when the native instance and its BindingContext is destroyed and OnBindingContextDestructedCallback is invoked."

File: wrappers/src/shared_realm_cs.cpp [hunk 4]
```diff
@@ -43,13 +43,19 @@ namespace binding {
     void (*s_open_realm_callback)(void* task_completion_source, ThreadSafeReference* ref, int32_t error_code, const char* message, size_t message_len);
     void (*s_realm_changed)(void* managed_state_handle);
     void (*s_get_native_schema)(SchemaForMarshaling schema, void* managed_callback);
+    void (*s_free_gc_handle)(void* managed_handle);
```
Comment (nirinchev) [inline wrappers/src/shared_realm_cs.cpp:46]: This variable should be renamed to loosely match the C# name.

**Review by nirinchev — APPROVED** [review]
Two nits from me, looks good otherwise.

File: Realm/Realm/Realm.cs (hunk 3 above)
Comment (papafe) [reply Realm/Realm/Realm.cs:201]: Agreed, it should have been "A realm is not responsible"

**Comment by papafe** [comment]
Corrected according to @nirinchev suggestions. I made the `Justification` for the `SuppressMessage` shorter, otherwise it feels more like a full fledged comment in my opinion

**Review by nirinchev — APPROVED** [review]
